Add validation for metrics config during parquet write

holdenk opened the pull request:
Validate that the metrics config refers to actual columns in the schema during write.

This is for issue #1548.

The discussion in the ticket suggested calling it during table creation, but I moved it to parquet write since we don't parse the metrics config until the write stage not in the create call. Still new to the code base so if that isn't the tradeoff we want hella ok to refactor.

rdblue commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -70,10 +71,17 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
+    for (String column : columnModes.keySet()) {
+      if (schema.findField(column) == null) {
+        throw new ValidationException("Could not find column %s from metrics in schema %s", column, schema);
+      }
```
You can use `ValidationException.check` like a `Precondition` to avoid adding lots of `if` statements:

```java
ValidationException.check(schema.findField(column) != null, "Could not find ...", ...);
```

holdenk replied:
Oh nice :)

rdblue commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -70,10 +71,17 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
           }
           spec.columnModes.put(columnAlias, mode);
         });
-
```
Nit: this is an unnecessary whitespace change that could lead to git conflicts. Could you add it back?

rdblue commented on parquet/src/main/java/org/apache/iceberg/parquet/Parquet.java:
```diff
@@ -226,6 +226,9 @@ private CompressionCodecName codec() {
       MessageType type = ParquetSchemaUtil.convert(schema, name);
 
+      // Check that our metrics make sense
+      metricsConfig.validateProperties(schema);
```
Is this the right place to do the validation?

If a user adds a bad property or performs some schema update that causes a validation error, that would break all writes to the table. To me, it doesn't seem like we are catching the problem early enough and possibly allowing a typo to break scheduled jobs.

What do you think about adding this validation when altering the table? `UpdateProperties` could check whether any properties starting with `write.metadata.metrics` were modified and run this. Similarly, `UpdateSchema` could run this as well, although I think that we should probably modify `UpdateSchema` to simply update the properties for column renames (if that's easily done).

holdenk replied:
Makes sense, I'll move this over to UpdateProperties/UpdateSchema tomorrow :)

aokolnychyi replied:
I'd also cover table creation. I feel that we should fail table creation if the metrics config refers to non-existing columns. I have seen so many cases when people clone an example but forget to update the metrics config and they realize the problem only after ingestion is done.

rdblue replied:
That's a good point. Table creation is a good time to validate this as well.

holdenk replied:
I've put this in for table creation & schema + property evolution. Let me know if I hooked in the correct places.

rdblue commented:
Thanks for working on this, @holdenk! It would be great to get this gap fixed.

holdenk commented:
Thanks for the quick review :) I'll circle back tomorrow with the changes :)

holdenk commented:
I think I've addressed the comments, although it's possible the hooks I've chosen are not the best places but it seems to cover the right areas with the tests. Let me know if there is a better place to put the hooks for validation.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -303,6 +304,9 @@ public Table create() {
 
       Map<String, String> properties = propertiesBuilder.build();
+      // Validate the metrics
+      MetricsConfig.fromProperties(properties).validateProperties(schema);
```
I think we need this in the `createTransaction` and other methods as well.

holdenk replied:
Oh cool, I didn't think of putting it there, I'll give that a shot.

aokolnychyi replied:
@rdblue, will moving the check to the constructor require us to also update the metrics config when the schema evolves? Otherwise, drop/rename operations will always fail if the metrics config refers to them?

aokolnychyi replied:
I think it is a good thing to update the metrics config implicitly when the schema evolves, BTW.

holdenk replied:
I think maybe automatically updating the metrics config on schema evolution could be a good follow up ticket, WDYT @aokolnychyi ?
I think for this first part requiring that the metrics config is valid during schema evolution is a good first step.

rdblue replied:
I agree that it is a follow-up. This will catch the problem and we can rewrite later.

We may also want to consider alternative storage for these that uses ID rather than name. I guess we could rewrite to ID-based.

aokolnychyi replied:
It is absolutely a follow-up item. Just wanted to bring attention to that.

holdenk replied:
Sounds good. I've added https://github.com/apache/iceberg/issues/2079

aokolnychyi replied:
Thanks, @holdenk!

rdblue replied:
After thinking about this a bit more, I don't think it is a good idea to add the check in the `TableMetadata` constructor after all. The problem with doing that is that it would break existing tables because they would validate when loading a table. Instead, we need to validate each case where a table is created so we could put it in the `newTableMetadata` factory method instead.

That would mean we also need to keep the validation in the property and schema update operations as you already have in this PR.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -74,6 +75,14 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
     return spec;
   }
 
+  public void validateProperties(Schema schema) {
```
I am not sure `validateProperties` is the most descriptive name in this case.

Something like `validateReferencedColumns` or similar would probably convey the purpose more. For example, we don't validate the property values here (if a value is invalid, it will be defaulted). We just check the config refers to valid columns.

Alternatively to changing the name, we can make this method static and accept both properties as well as schema. For example, we may invoke it as `MetricsConfig.validateProperties(props, schema)` and it will check both the column names as well as metrics modes (we can overload `fromProperties` so that it will throw an exception instead of defaulting).

Thoughts, @holdenk @rdblue @RussellSpitzer @shardulm94?

aokolnychyi replied:
I'd probably go for a static method and validating the value modes as well.

holdenk replied:
I'm not sure about having this be a static method since it would end up constructing an instance of the object anyways, but I don't have strong feelings here. I lean towards giving it a name like `validateReferencedColumns` personally.

aokolnychyi replied:
Same for me. I don't feel strongly about this one.

rdblue replied:
I think it is fine either way. I think `MetricsConfig#validateProperties` is reasonable, but I like `validateReferencedColumns` as well.

aokolnychyi commented on core/src/main/java/org/apache/iceberg/PropertiesUpdate.java:
```diff
@@ -89,6 +89,11 @@ public UpdateProperties defaultFormat(FileFormat format) {
 
+    // Validate the metrics
+    if (base.schema() != null) {
+      MetricsConfig.fromProperties(newProperties).validateProperties(base.schema());
```
If we go for a static method, this will also become shorter:

```
MetricsConfig.validateProperties(newProperties, base.schema());
```

aokolnychyi commented:
@rdblue @shardulm94 @RussellSpitzer, not related to this PR but a general question: the metrics config relies on column names instead of column ids as it is controlled through table properties. I was wondering whether it is safe to do so.

I considered the following cases:
- While writing new data, it should be always OK to follow the name-based approach if the config is up-to-date.
- While importing data (assuming we have a correct name mapping), we will assign correct ids to columns, use the ids to find the current aliases in the schema, so this should work.
- When fixing (re-importing) metadata (assuming we have renamed a column through Iceberg and no name mapping), we will just use the column id in the file to find the current name, so this should work.

Overall, it seems safe to me. Any cases I missed?

rdblue commented on core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:
```diff
@@ -155,6 +156,42 @@ public void testDeleteColumn() {
+    NameMapping updated = NameMappingParser.fromJson(updatedJson);
+
+    // should not change the mapping
+    validateUnchanged(mapping, updated);
```
It seems strange to me that this would be tested in the suite for the schema mapping. This should be independent of mapping. While it is okay to have a test that a mapping is not updated if the validation fails, that should be covered by not updating the schema if metrics property validation fails. Could you start a new suite similar to `TestSchemaAndMappingUpdate` for this instead of adding the tests here?

holdenk replied:
Sure I can make a new suite for this.

holdenk replied:
Sure I'll add a new test suite next.

kbendick commented on core/src/test/java/org/apache/iceberg/TestSchemaAndMappingUpdate.java:
```diff
@@ -155,6 +156,42 @@ public void testDeleteColumn() {
+        null,
+        () ->
+          table.updateSchema()
+              .deleteColumn("id")
+          .commit());
```
Nit: line indentation doesn't line up with the other values.

rdblue commented on core/src/main/java/org/apache/iceberg/TableMetadata.java:
```diff
@@ -110,6 +110,11 @@ static TableMetadata newTableMetadata(Schema schema,
+    // break existing tables.
+    MetricsConfig.fromProperties(properties).validateReferencedColumns(schema);
+
+
```
Nit: double newline.

rdblue commented on core/src/main/java/org/apache/iceberg/hadoop/HadoopTables.java:
```diff
@@ -303,7 +303,8 @@ public Table create() {
       Map<String, String> properties = propertiesBuilder.build();
-      TableMetadata metadata = tableMetadata(schema, spec, sortOrder, properties, location);
+
+      TableMetadata metadata = TableMetadata.newTableMetadata(schema, spec, sortOrder, location, properties);
```
Is it necessary to change this file? Doesn't `tableMetadata` call `newTableMetadata`?

holdenk replied:
My bad, I'll revert this change.

rdblue commented on spark/src/test/java/org/apache/iceberg/spark/source/TestWriteMetricsConfig.java:
```diff
@@ -228,6 +230,23 @@ public void testCustomMetricCollectionForParquet() throws IOException {
+        () -> tables.create(SIMPLE_SCHEMA, spec, properties, tableLocation));
+  }
+
+
```
Nit: extra newline.

rdblue commented on core/src/main/java/org/apache/iceberg/PropertiesUpdate.java:
```diff
@@ -89,6 +89,11 @@ public UpdateProperties defaultFormat(FileFormat format) {
 
+    // Validate the metrics
+    if (base != null && base.schema() != null) {
+      MetricsConfig.fromProperties(newProperties).validateReferencedColumns(base.schema());
```
I don't think this is tested. Could you add a test to the property update tests?

holdenk replied:
This is covered in the same test as the other one (the exception is on a different line though):

```
org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 1] FAILED
    java.lang.AssertionError: No exception was thrown (Creating metrics for non-existent column fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:169)

org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 2] FAILED
    java.lang.AssertionError: No exception was thrown (Creating metrics for non-existent column fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:169)

```

rdblue commented on core/src/main/java/org/apache/iceberg/SchemaUpdate.java:
```diff
@@ -357,7 +357,14 @@ private void internalMove(String name, Move move) {
+      MetricsConfig.fromProperties(base.properties()).validateReferencedColumns(newSchema);
+    }
+
+    return newSchema;
```
I don't think that this path is tested. Could you add one?

holdenk replied:
It is currently covered by `testModificationWithMetricsMetrics`, commenting out the check results in the following error when running gradle build:

```
org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 1] FAILED
    java.lang.AssertionError: No exception was thrown (Deleting a column with metrics fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:179)

org.apache.iceberg.TestSchemaAndMappingUpdate > testModificationWithMetricsMetrics[formatVersion = 2] FAILED
    java.lang.AssertionError: No exception was thrown (Deleting a column with metrics fails), expected: org.apache.iceberg.exceptions.ValidationException
        at org.junit.Assert.fail(Assert.java:88)
        at org.apache.iceberg.AssertHelpers.assertThrows(AssertHelpers.java:65)
        at org.apache.iceberg.TestSchemaAndMappingUpdate.testModificationWithMetricsMetrics(TestSchemaAndMappingUpdate.java:179)

```

aokolnychyi commented on core/src/main/java/org/apache/iceberg/MetricsConfig.java:
```diff
@@ -74,6 +75,14 @@ public static MetricsConfig fromProperties(Map<String, String> props) {
+    for (String column : columnModes.keySet()) {
+      ValidationException.check(
+          schema.findField(column) != null,
+          "Could not find column %s from metrics in schema %s", column, schema);
```
Just thinking out loud: will be clear for an inexperienced user what is wrong?
Is `from metrics` descriptive enough? Should we make it `from metrics config`?
Or make it as `Invalid metrics config: could not find column %s in schema %s`?

rdblue replied:
That seems like a fair criticism. What about adding the bad property name and noting that it is a table property? That would help.

aokolnychyi commented:
LGTM apart from the existing comments.

rdblue commented:
Thanks, @holdenk! Really nice to have this done.
